### ISSUE TITLE
Error [ERR_UNKNOWN_BUILTIN_MODULE]: No such built-in module: express

### ISSUE THREAD

**Issue by MGMehdi** [nodejs/single-executable#87]
Hello,
I tried to compile my express app following the `Single executable applications` guide in but I've got this error :
```sh
node:internal/util/embedding:48
    throw new ERR_UNKNOWN_BUILTIN_MODULE(id);
    ^

Error [ERR_UNKNOWN_BUILTIN_MODULE]: No such built-in module: express
    at embedderRequire (node:internal/util/embedding:48:11)
    at dist/app.js:141:33
    at embedderRunCjs (node:internal/util/embedding:37:10)
    at node:internal/main/embedding:18:8 {
  code: 'ERR_UNKNOWN_BUILTIN_MODULE'
}

Node.js v21.6.0
```
It also appear with Node.js v20.10.0.

The app is written with typescript but the build is working when running in the terminal. Here is the `tsconfig.json`
```json
{
  "compilerOptions": {
    "target": "ES5",
    "lib": ["ES5"],
    "module": "commonjs",
    "rootDir": "./src",
    "sourceMap": true,
    "outDir": "./dist",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```
Did I miss something ?

**Comment by joyeecheung** [comment]
Yes, for now you would need to bundle your code into one single file before injecting it into the SEA.

**Comment by MGMehdi** [comment]
Hooo I didn't understand like that. Thanks for the clarification! 
For now if I want to achieve the same results is by using `pkg` by vercel?

**Comment by robertsLando** [comment]
FYI pkg has been archived now. I'm maintaining the most active fork here: https://github.com/yao-pkg/pkg

If you need ESM support there some alternatives: https://github.com/yao-pkg/pkg/issues/16#issuecomment-1945486658

**Comment by antonioacag** [comment]
> Yeah, the part where it says that it [only supports built-in modules](https://nodejs.org/api/single-executable-applications.html#requireid-in-the-injected-module-is-not-file-based).
Someone be better deleting this type of unnecessary answers.

**Comment by pironev** [comment]
This will works if you do a build with just one file.
Install esbuild
    `npm install --save-exact --save-dev esbuild`
Copy to its package.json
     **"scripts": {
     "build": "esbuild controllerServer.js --bundle --platform=node --outfile=out.js"
   }**
generate out.js :
    **npm run build**
Then it resumes the whole process from https://nodejs.org/api/single-executable-applications.html

**Comment by Cycov** [comment]
> **This will works if you do a build with just one file.** Install esbuild: `npm install --save-exact --save-dev esbuild` Copy to package.json:
> 
> ```
>      "scripts": {
>      "build": "esbuild your-start-file.js --bundle --platform=node --outfile=out.js"
>         }
> ```
> 
> generate out.js : ` npm run build` Then resumes the whole process from https://nodejs.org/api/single-executable-applications.html for out.js

In my case I need to package a application that uses the `serialport` module. The compiled version throws the following error:
`Error: No native build was found for platform=win32 arch=x64 runtime=node abi=127 uv=1 libc=glibc node=22.2.0`

**Comment by robertsLando** [comment]
@Cycov check https://github.com/yao-pkg/pkg/issues/16#issuecomment-1945486658

**Comment by BillChen-AI** [comment]
> **This will works if you do a build with just one file.** Install esbuild: `npm install --save-exact --save-dev esbuild` Copy to package.json:
> 
> ```
>      "scripts": {
>      "build": "esbuild your-start-file.js --bundle --platform=node --outfile=out.js"
>         }
> ```
> 
> generate out.js : ` npm run build` Then resumes the whole process from https://nodejs.org/api/single-executable-applications.html for out.js

I followed the steps you provided and successfully resolved a `fast-csv` built-in module not found issue. 
The generated exe runs without an error on Windows 10 64-bit.

**Comment by codenoid** [comment]
@robertsLando I've tried your esbuild.js file, but it's not works for me :(, I'm having issue with `usb` package

**Comment by cspotcode** [comment]
Native modules (DLLs) cannot be bundled by esbuild, so anyone trying to use a native module will hit trouble.

For everyone else, you may have luck using an all-in-one solution:

```
$ deno compile -A --output ./outputexecutable ./entrypoint.ts
```

This should emit an executable file with all dependencies bundled into it without needing a separate `esbuild` step.

**Comment by iamstarkov** [comment]
@cspotcode wild of you to throw Deno in node discussions =)

**Comment by MGMehdi** [comment]
@cspotcode, using Deno to create an executable work but you can open it with a notepad and see all the code.
I've move to Golang for that reason

**Comment by cspotcode** [comment]
Based on these comments, my suggestion is, indeed, a solution to this issue:
You can take a node codebase and compile it to a single executable using `deno compile`, without additional fanfare.

https://github.com/nodejs/single-executable/issues/87#issuecomment-2651810469

**Comment by codenoid** [comment]
I'm rewriting my app to Rust instead for cross compilation

**Comment by MGMehdi** [comment]
@cspotcode and with a visible codebase via a notepad when you open the .exe for example with that
